gapis/api/vulkan/framegraph.go

```diff
@@ -166,6 +212,45 @@ type framegraphInfoHelpers struct {
 	rpInfos  []*renderpassInfo
 	rpInfo   *renderpassInfo
 	currRpId uint64
+	// imageLookup is a lookup table to quickly find images that match a memory
+	// access. Scanning the whole state to look for an image matching a memory
+	// access is slow, this lookup table saves seconds of computation. It is
+	// updated when a new renderpass is started under a new top-level parent
+	// command: images can only be created/destroyed between top-level commands,
+	// not during the execution of subcommands.
+	imageLookup  map[memory.PoolID]map[memory.Range][]*ImageObjectʳ
+	parentCmdIdx uint64 // used to know when to update imageLookup
+}
+
+// updateImageLookup updates the lookup table to quickly find an image matching a memory observation.
+func (helpers *framegraphInfoHelpers) updateImageLookup(state *State) {
+	helpers.imageLookup = make(map[memory.PoolID]map[memory.Range][]*ImageObjectʳ)
+	for _, image := range state.Images().All() {
+		for _, aspect := range image.Aspects().All() {
+			for _, layer := range aspect.Layers().All() {
+				for _, level := range layer.Levels().All() {
+					pool := level.Data().Pool()
+					memRange := level.Data().Range()
+					if _, ok := helpers.imageLookup[pool]; !ok {
+						helpers.imageLookup[pool] = make(map[memory.Range][]*ImageObjectʳ)
+					}
+					imgObj := image
+					helpers.imageLookup[pool][memRange] = append(helpers.imageLookup[pool][memRange], &imgObj)
```

Comment:
We haven't check if the range is already existing or any existing range includes the current range. If I understand correctly, what happens here behind the scene is;

- check if memRange exist for the map stored imageLookup[pool] 
- if exists add &imgObj to the slice stored in imageLookup[pool][memRange]
- if not, then append imgObj to a nil slice(which creates it) then assign that slice to imageLookup[pool][memRange] which again allocates some stuff on the map.

If this is the intended behaviour then it's ok. Otherwise, please reconsider.

Reply:
You got the idea, we append to helpers.imageLookup[pool][memRange], which may be nil.

This seems to be the classic way to add items to a slice member of a map, quoting https://blog.golang.org/maps:

> Another instance of helpful zero values is a map of slices. Appending to a nil slice just allocates a new slice, so it's a one-liner to append a value to a map of slices; there's no need to check if the key exists.